File: C#/Memento/Caretaker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Memento.Memento
{
    class Caretaker
    {
        private List<IMemento> history;
        private IOriginator originator;

        public Caretaker(IOriginator originator)
        {
            Console.WriteLine("CARETAKER: Creating caretaker");
            this.originator = originator;
            history = new List<IMemento>();
        }

        //adding memento to the history
        public void Backup()
        {
            Console.WriteLine("CARETAKER: Adding memento to history");
            history.Add(originator.CreateBackup());
        }

        //restoring originator state from last element in history
        public void Restore()
        {
            if (history.Count > 0 && originator != null)
            {
                Console.WriteLine("CARETAKER: Restoring from history");
                history.Last().RestoreMemento();
                history.Remove(history.Last());
            }
        }

        public void ShowHistory()
        {
            Console.WriteLine("HISTORY");
            foreach (IMemento mem in history)
            {
                Console.WriteLine(mem.ToString());
            }
        }

    }
}

File: python/Memento/Caretaker.py
from IMemento import IMemento
from IOriginator import IOriginator

class Caretaker:

    def __init__(self, originator: IOriginator):
        print("CARETAKER: Creating cretaker")
        self.originator = originator
        self.history = list([IMemento])
        self.history.clear()

    def Backup(self) -> None:
        print("CARETAKER: Adding memento to history")
        self.history.append(self.originator.CreateBackup())

    def Restore(self) -> None:
        if (len(self.history) > 0):
            print("CARETAKER: Restoring from history")
            self.history[-1].RestoreMemento()
            self.history.pop()
    
    def ShowHistory(self) -> None:
        print("\nHISTORY")                      
        for mem in self.history:
            print(mem.ShowMemento())